src/services/navigator/helpers.ts

```diff
@@ -92,3 +93,131 @@ export const getUrlFromHref = (
+export const validateUrl = (
+  action: TypesLegacy.NavAction,
+  routeParams: TypesLegacy.NavigationRouteParams,
+) => {
+  if (action === TypesLegacy.NAV_ACTIONS.PUSH || TypesLegacy.NAV_ACTIONS.NEW) {
+    if (!routeParams.url || !cleanHrefFragment(routeParams.url)) {
+      throw new Errors.HvNavigatorError(
+        `Route params must include a url for action '${action}'`,
+      );
+    }
+  }
+};
+
+/**
+ * Recurse through the route states to find the target
+ * If the target is found, the path is generated to the first element
+ * example: ['home', 'shifts', 'my-shifts']
+ */
+export const findPath = (
+  state: Imports.NavigationState,
+  targetId: string,
+): string[] => {
+  let path: string[] = [];
+  if (!state) {
+    return path;
+  }
+
+  const { routes } = state;
+  if (!routes) {
+    return path;
+  }
+  routes.every(route => {
+    if (route.name === targetId) {
+      path.unshift(route.name);
+      return false;
+    }
+    path = [
+      ...path,
+      ...findPath(route.state as Imports.NavigationState, targetId),
+    ];
+    // If the recursion found the target, add the current route name to the path as we back out
+    if (path.length) {
+      path.unshift(route.name);
+      return false;
+    }
+    return true;
+  });
+  return path;
+};
+
+/**
+ * Continue up the hierarchy until a navigation is found which contains the target
+ * If the target is not found, no navigation is returned
+ * If no target is provided, the current navigation is returned
+ */
+export const getNavigatorAndPath = (
+  targetId?: string,
+  navigation?: HvRoute.RNTypedNavigationProps,
+): [HvRoute.RNTypedNavigationProps?, string[]?] => {
+  if (!targetId) {
+    return [navigation, undefined];
+  }
+  if (navigation) {
+    const path = findPath(navigation.getState(), targetId);
+    if (path.length) {
+      return [navigation, path];
+    }
+    return getNavigatorAndPath(targetId, navigation.getParent());
+  }
+  return [undefined, undefined];
+};
+
+/**
+ * Generate a nested param hierarchy with instructions for each screen
+ * to step through to the target
+ * example: { screen: 'home', params:
+ *    { screen: 'shifts', params:
+ *        { screen: 'my-shifts', params:
+ *            { url: 'someurl.xml' } } } }
+ */
+export const buildParams = (
+  routeId: string,
+  path: string[],
+  routeParams: TypesLegacy.NavigationRouteParams,
+  index = 0,
+): Types.NavigationNavigateParams | TypesLegacy.NavigationRouteParams => {
+  let param: Types.NavigationNavigateParams;
+  if (path.length && index < path.length) {
+    const screen = path[index];
+
+    if (!screen) {
+      throw new Errors.HvNavigatorError('screen is undefined');
+    }
+    param = { screen };
+    param.params = buildParams(routeId, path, routeParams, index + 1);
+  } else {
+    param = { screen: routeId };
+    // The last screen in the path receives the route params
+    // example: { url: 'someurl.xml' }
+    param.params = routeParams;
+  }
+  return param;
```

Comment:
minor: remove mutation

```suggestion
  if (path.length && index < path.length) {
    const screen = path[index];

    if (!screen) {
      throw new Errors.HvNavigatorError('screen is undefined');
    }
    return {
      screen,
      params: buildParams(routeId, path, routeParams, index + 1);
    }
  } else {
    return {
      screen: routeId,
      // The last screen in the path receives the route params
      // example: { url: 'someurl.xml' }
      params: routeParams,
    };
  }
```

Reply:
I've replaced this and several other spots where mutation was occurring